src/megatron/energon/task_encoder/base.py

```diff
@@ -100,6 +109,125 @@ def batch_list(batch: List[Any]) -> Any:
+    if restore_seeds:
+        worker_seed = None
+
+        @functools.wraps(fn)
+        def seed_wrapper_generator(self, *args, **kwargs):
+            nonlocal worker_seed
+            if worker_seed is None:
+                worker_seed = WorkerConfig.active_worker_config.worker_seed()
+
+            # Use a deterministic hash function to compute the seed
+            hash_digest = hashlib.sha1(
+                f"{worker_seed}|{self.current_sample_index}".encode("utf-8")
+            ).digest()
+
+            # We use the first 4 bytes of the hash as the seed and fix the endianness
+            inner_seed_value = int.from_bytes(hash_digest[:4], byteorder="big")
+
+            # Save the RNG states and set the new seed
+            outer_rng_state = save_global_rng_state()
+                    
+            # Before constructing the generator and before the first
+            # iteration, set inner RNG based on computed seed
+            set_global_seeds(inner_seed_value)
+            
+            it = iter(fn)
+            
+            inner_rand_state = None
+
+            while True:
+                
+                if inner_rand_state is not None:
+                    # Restore inner random state before calling the generator
+                    # This will not be done on the first iteration
+                    restore_global_rng_state(inner_rand_state)
+                
+                try:
+                    # Now call the generator. This will yield the sample
+                    # But note it may also throw an exception or a StopIteration
+                    sample = next(it)
+                    
+                    # Save inner random state after calling the generator
+                    inner_rand_state = save_global_rng_state()
+                except StopIteration:
+                    # We're stopping here, but the outer random state
+                    # will be restored before returning (in finally below)
+                    break
+                finally:                                            
+                    # Restore outer rand state before yielding or when an exception was raised
+                    restore_global_rng_state(outer_rng_state)
+                
+                # Now yield the sample.
+                # This will give control back to the caller who may
+                # change the random state.
+                yield sample
+                
+                # Save outer random state after yielding
+                outer_rng_state = save_global_rng_state()
+        
+                
+
+        @functools.wraps(fn)
+        def seed_wrapper(self, *args, **kwargs):
+            nonlocal worker_seed
+            if worker_seed is None:
+                worker_seed = WorkerConfig.active_worker_config.worker_seed()
+
+            # Use a deterministic hash function to compute the seed
+            hash_digest = hashlib.sha1(
```

Comment:
Can this also be moved to the seeding code? I believe we have it in more places, e.g. where we initialize the seeds